### PR TITLE
BNGP-1477 Application Reference

### PR DESCRIPTION
**Pull request by teddmason**

### CODE REVIEW

**Comment by teddmason** [comment]
This PR includes changes made on https://github.com/DEFRA/biodiversity-net-gain-service/pull/218

**Comment by teddmason** [comment]
@pwadmore-ea I can take you through the logic for this at some point.  Key risks that i thought of that I'm hoping the work addresses are as follows:

- Generating a duplicate reference number for multiple users submitting at the same time
- reference ordering and counting correctly when we clock change to BST

**Review by pwadmore-ea — CHANGES_REQUESTED** [review]

File: packages/database-version-control/src/migrations/2023.01.04T10.57.15.create-bng-schema.sql [hunk 1]
```diff
@@ -0,0 +1,14 @@
+DO
+$$
+BEGIN
+  IF NOT EXISTS (select nspname from pg_catalog.pg_namespace where nspname = 'bng') THEN
+     CREATE SCHEMA bng;
```
[the remark this answers is not in the document]
Comment (teddmason) [reply packages/database-version-control/src/migrations/2023.01.04T10.57.15.create-bng-schema.sql:5]: I'd rather not, as it wraps the USAGE and PRIVALEDGE statements too

File: packages/database-version-control/src/migrations/2023.01.04T15.26.58.create-application-reference-function.sql [hunk 2]
```diff
@@ -0,0 +1,43 @@
+create function bng.fn_create_application_reference()
```
[the remark this answers is not in the document]
Comment (teddmason) [reply packages/database-version-control/src/migrations/2023.01.04T15.26.58.create-application-reference-function.sql:1]: CREATE FUNCTION IF NOT EXISTS is not valid.  Only option would be to have a drop function before the create.  But I see the need for the IF exists on the source control to be minor. If the NOT EXISTS checks are required then something has gone wrong with the database setup, and potentially having the NOT EXISTS checks could hide that

**Comment by pwadmore-ea** [comment]
> @pwadmore-ea I can take you through the logic for this at some point. Key risks that i thought of that I'm hoping the work addresses are as follows:
> 
>     * Generating a duplicate reference number for multiple users submitting at the same time
> 
>     * reference ordering and counting correctly when we clock change to BST

@teddmason - Hopefully this comment reflects what we have discussed outside GitHub:

- Addition of a unique constraint on the application_reference column of the database table to guard against allocation of duplicate references.
- Use of timestamp with timezone column on the database table with UTC as the time zone so no local times are stored in the database. I acknowledge that this means some allocations during BST are at risk of being assigned to the wrong day. If this causes issues, potential solutions can be investigated accordingly.

**Review by pwadmore-ea — APPROVED** [review]

**Pull request merged**